File: 王晓东习题程序/1-2.cpp
#include <iostream>
#include <cstring>
#include <fstream>

using namespace std;

int C(int, int);
int N(char [], int);

int main() 
{
    ifstream input("data\\input1_2.txt");
    ofstream outfile("data/output1_2.txt");
    if (!input) cout << "文件打开错误!" << endl;
    else
    {
        int k=10;
        // input >> k;
        for (int m=0; m<=k; m++)
        {
            char s[7];
            input.getline(s, 7);
            if (!s[0]) continue;
            int n = strlen(s);
            int sum = 0;
            for(int j=1;j<n;j++)
            {
                sum += C(26, j);
            }
            int result = sum + N(s, 0) + 1;
            cout << result << endl;
            outfile << result << endl;
        }
    }
    
    input.close();
    outfile.close();
    return 0;
}

int C(int n, int m)
{
    int a = 1;
    int b = 1;
    for(int j=1;j<=m;j++)
    {
        a *= n-j+1;
        b *= j;
    }
    return a/b;
}

int N(char *s, int i)
{
    int n = strlen(s);
    if(n<1) return 0;
    char high = s[0];
    int sum = 0;
    for(int k=1;k<=high-'a'-i;k++)
    {
        sum += C(26-i-k, n-1);
    }
    if(n>1) sum += N(&s[1], high-'a'+1);
    return sum;
}

File: 王晓东习题程序/1-2.py
'''
    字典序问题，求出字典序对应的编码
'''


def C(n: int, m: int) -> int:
    """求C_n^m
    """
    a, b = 1, 1
    for i in range(1, m+1):
        a *= n-i+1
        b *= i
    return int(a / b)


def N(s: str, i: int) -> sum:
    """ 计算字符长度为n时，字符串的编号增量
         s:字符串，i为确定首字母时要排除的字母个数
    """
    sum = 0
    n = len(s)
    if n < 1:
        return 0
    high = ord(s[0])
    for k in range(1, high-ord('a')-i+1):  # 按字母顺序遍历字符串第一位到a+(i)位
        sum += C(26-k-i, n-1)  # 计算这些字母带来的编号增量
    if n > 1:
        sum += N(s[1:], high-ord('a')+1)  # 递归计算s[1:]带来的编号增量
    return sum


if __name__ == '__main__':
    with open("./data/input1_2.txt", 'r') as f:
        slist = f.readlines()
    with open("./data/output1_2.txt", "a") as f:
        for s in slist:
            s = s.strip().replace("\n", "")
            sum = 0
            n = len(s)
            for j in range(n):  # 计算长度小于n的字典序的最大编号
                sum += C(26, j)
            result = sum + N(s, 0)
            print(s)
            print(result)

            f.write("%d\n" % result)
    # s = "abcd"
    # sum = 0
    # n = len(s)
    # for j in range(1, n): #计算长度小于n的字典序的最大编号
    #     sum += C(26, j)
    # result = sum + N(s, 0) + 1
    # print(result)
